kong/clustering/rpc/socket.lua

```diff
@@ -0,0 +1,287 @@
+local cjson_decode = cjson.decode
+local is_timeout = utils.is_timeout
+local unpack = table.unpack
+local exiting = ngx.worker.exiting
+local ngx_time = ngx.time
+local ngx_log = ngx.log
+
+
+local CLUSTERING_PING_INTERVAL = constants.CLUSTERING_PING_INTERVAL
+local PING_WAIT = CLUSTERING_PING_INTERVAL * 1.5
+local PING_TYPE = "PING"
+local PONG_TYPE = "PONG"
+local ngx_WARN = ngx.WARN
+local ngx_DEBUG = ngx.DEBUG
+
+
+-- create a new socket wrapper, wb is the WebSocket object to use
+-- timeout and max_payload_len must already been set by caller when
+-- creating the `wb` object
+function _M.new(manager, wb, node_id)
+  local self = {
+    wb = wb,
+    interest = {}, -- id: callback pair
+    outgoing = queue.new(4096), -- write queue
+    manager = manager,
+    node_id = node_id,
+    sequence = 0,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:_get_next_id()
+  local res = self.sequence
+  self.sequence = res + 1
+
+  return res
+end
+
+
+-- start reader and writer thread and event loop
+function _M:start()
+  self.read_thread = ngx.thread.spawn(function()
+    local last_seen = ngx_time()
+
+    while not exiting() do
+      local data, typ, err = self.wb:recv_frame()
+
+      if err then
+        if not is_timeout(err) then
+          return nil, err
+        end
+
+        local waited = ngx_time() - last_seen
+        if waited > CLUSTERING_PING_INTERVAL then
+          local res, err = self.outgoing:push(PING_TYPE)
+          if not res then
+            return nil, "unable to send ping: " .. err
+          end
+        end
+
+        if waited > PING_WAIT then
+          return nil, "did not receive ping frame from other end within " ..
+                      PING_WAIT .. " seconds"
+        end
+
+        -- timeout
+        goto continue
+      end
+
+      last_seen = ngx_time()
+
+      if typ == "ping" then
+        local res, err = self.outgoing:push(PONG_TYPE)
+        if not res then
+          return nil, "unable to handle ping: " .. err
+        end
+
+        goto continue
+      end
+
+      if typ == "pong" then
+        ngx_log(ngx_DEBUG, "[rpc] got PONG frame")
+
+        goto continue
+      end
+
+      if typ == "close" then
+        return true
+      end
+
+      assert(typ == "binary")
+
+      local payload = cjson_decode(data)
+      assert(payload.jsonrpc == "2.0")
+
+      if payload.method then
+        -- invoke
+
+        local cb = self.manager.callbacks.callbacks[payload.method]
+        if not cb then
+          local res, err = self.outgoing:push({
+            jsonrpc = "2.0",
+            id = payload.id,
+            ["error"] = {
+              code = jsonrpc.METHOD_NOT_FOUND,
+              message = "Method not found",
+            }
+          })
+          if not res then
+            return nil, "unable to handle ping: " .. err
+          end
+
+          goto continue
+        end
+
+        -- call dispatch
+
+        assert(ngx.timer.at(0, function(premature)
```

Comment:
could we turn this closure to a local function like `dispath(premature, self, payload)`?